pkg/manager/impl/reservation_manager.go

```diff
@@ -0,0 +1,189 @@
+}
+
+func NewReservationManager(
+	repo repositories.RepositoryInterface,
+	reservationTimeout time.Duration,
+	nowFunc NowFunc, // Easier to mock time.Time for testing
+	reservationScope promutils.Scope,
+) interfaces.ReservationManager {
+	systemMetrics := reservationMetrics{
+		scope: reservationScope,
+		reservationAcquired: labeled.NewCounter(
+			"reservation_acquired",
+			"Number of times a reservation was acquired",
+			reservationScope),
+		reservationAlreadyInProgress: labeled.NewCounter(
+			"reservation_already_in_progress",
+			"Number of times we try of acquire a reservation but the reservation is in progress",
+			reservationScope,
+		),
+		acquireReservationFailure: labeled.NewCounter(
+			"acquire_reservation_failure",
+			"Number of times we failed to acquire reservation",
+			reservationScope,
+		),
+		getTagFailure: labeled.NewCounter(
+			"get_tag_failure",
+			"Number of times we failed to get tag",
+			reservationScope,
+		),
+	}
+
+	return &reservationManager{
+		repo:               repo,
+		reservationTimeout: reservationTimeout,
+		now:                nowFunc,
+		systemMetrics:      systemMetrics,
+	}
+}
+
+func (r *reservationManager) GetOrReserveArtifact(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (*datacatalog.GetOrReserveArtifactResponse, error) {
+	tagKey := transformers.ToTagKey(request.DatasetId, request.TagName)
+	tag, err := r.repo.TagRepo().Get(ctx, tagKey)
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Tag does not exist yet, let's acquire the reservation to work on
+			// generating the artifact.
+			status, err := r.tryAcquireReservation(ctx, request)
+			if err != nil {
+				r.systemMetrics.acquireReservationFailure.Inc(ctx)
+				return nil, err
+			}
+
+			return &datacatalog.GetOrReserveArtifactResponse{
+				Value: &datacatalog.GetOrReserveArtifactResponse_ReservationStatus{
+					ReservationStatus: &status,
+				},
+			}, nil
+		}
+		logger.Errorf(ctx, "Failed retrieve tag: %+v, err: %v", tagKey, err)
+		r.systemMetrics.getTagFailure.Inc(ctx)
+		return nil, err
+	}
+
+	artifact, err := transformers.FromArtifactModel(tag.Artifact)
+	if err != nil {
+		return nil, err
+	}
+
+	return &datacatalog.GetOrReserveArtifactResponse{
+		Value: &datacatalog.GetOrReserveArtifactResponse_Artifact{
+			Artifact: artifact,
+		},
+	}, nil
+}
+
+func (r *reservationManager) tryAcquireReservation(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (datacatalog.ReservationStatus, error) {
+	repo := r.repo.ReservationRepo()
+	reservationKey := transformers.ToReservationKey(*request.DatasetId, request.TagName)
+	rsv, err := repo.Get(ctx, reservationKey)
+
+	resvExists := true
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Reservation does not exist yet so let's create one
+			resvExists = false
+		} else {
+			return datacatalog.ReservationStatus{}, err
+		}
+	}
+
+	now := r.now()
+	if !resvExists || rsv.ExpireAt.Before(now) {
```

Comment:
We need some epsilon value to update... maybe it should be configurable (next to `ReservationTimeout`) or maybe it should just be a multiplier of that... so basically the condition here should look something like this:

```suggestion
	if !resvExists || rsv.ExpireAt.Before(now.Add(r.reservationTimeoutBuffer)) {
```

Or:
```suggestion
	if !resvExists || rsv.ExpireAt.Before(now.Add(r.reservationTimeout / r.reservationTimeoutGracePeriodMultiplier)) {
```

Reply:
Can you explain a bit more on why we need `now.Add(r.reservationTimeout / r.reservationTimeoutGracePeriodMultiplier)`? We already store `now.Add(r.reservationTimeout)` to `expireAt` at the time we persist the reservation.

Reply:
I realized what I'm looking for, it should be called `heartbeatInterval` and it should do `now.Add(r.heartbeatInterval * 2)` or something... here is why:

t0: client requests reservation
t0+heartbeat: client requests reservation (heartbeat) and is returned the same reservation
t0+reseravtionTimeout-heartbeat: client requests reservation (heartbeat) and is returned the same reservation (since it hasn't expired yet)
t0+reservationTimeout+epsilon (since clocks are not synced): a different client requests the reservation and wins because it now has expired... right? and this is bad...

If, on the other side, you add this `heartbeatInterval*2` check, what will happen is:
t0: client requests reservation
t0+heartbeat: client requests reservation (heartbeat) and is returned the same reservation
t0+reseravtionTimeout-heartbeat: client requests reservation (heartbeat) and is returned an 'extended' reservation (since it'll be considered expired/about to expire)
t0+reservationTimeout+epsilon (since clocks are not synced): a different client requests the reservation, it'll lose because the original client still has the reservation

Reply:
Let me clarify a bit. The `reservation_timeout` is essentially a heartbeat timeout. Once acquired a reservation, the client is responsible to heartbeat by calling `ExtendReservation` API at an interval less than the `reservation_timeout` otherwise the reservation may expire and the other client can take over the reservation. Let's use an example to illustrate this and let's assume the `reservation_timeout` is 5 min.

| Time | What | expireAt |
| --- | --- | --- |
| 00:00 | A try to acquire reservation via GetOrReserveArtifact API and succeeded | 00:05 |
| 00:02 | A heartbeats via ExtendReservation and that reset expireAt to now (00:02)+reservation_timeout (5min) = 00:08  | 00:08 |
| 00:05 | B try to acquire reservation via GetOrReserveArtifact API and failed | 00:08 |
| 00:09 | C try to acquire reservation via GetOrReserveArtifact API and succeeded. A did not heartbeat again by 00:08 so reservation expired and C took over the reservation | -- |

Reply:
And if we set `reservation_timeout` to be 2-3X heartbeat interval, we could avoid the issue caused by clock not sync'd issue.

Reply:
aha, so that's what I was missing... for some reason I assumed GetOrReserve is the only available api and hence where the extension should happen...

Alright, so let's explore a startup scenario (after a deployment or crash... etc.).. the client starts up and it should immediately try to `ExtendReservation`, right? there doesn't seem to be a value in calling `GetOrReserve` in that case because even if the reservation is returned, it might expire as soon as it's received by the client...
So the only use-case for GetOrReserve is when it's the very first time it's reserving (or if it knows it lost the reservation)... after that it should only rely on ExtendReservation until it gets a response saying it lost the reservation...

I was thinking that even for the startup scenario, the client will call GetOrReserve (and not have to maintain state) and can reliably trust the response here... the way the code is done now, this can happen:
```
resp := GetOrReserve()
if resp.OwnerId == me {
    err := ExtendReservation
    👎🏾  it fails because in between the two calls, it expired and somebody else stole it...
}
```

if you do add the heartbeat interval to the check, GetOrReserve will be much more stable in that regard, right?

Another question RE: ExtendReservation, will it fail if the reservation has indeed expired but hasn't been claimed by anybody else? I guess it should just succeed, right?